Add reviewing guidelines

lindsayad opened the pull request:
@idaholab/moose-ccb let's iterate on this. Please all feel free to push additions/changes to this PR. I'll keep this in draft for now as we're iterating, and eventually obviously we'll need a link to this page

lindsayad committed:
Add reviewing guidelines

refs #000

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,34 @@
+
+## Review Guidelines
+
+- Contributors should be regarded as innocent until proven guilty
+  - Contributors probably aren't out to destroy MOOSE
+  - The contributor has probably thought about their contribution a lot
```
```suggestion
```
this was informal discussion

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,34 @@
+- Are there things we can do to abate fear of bugs?
+  - Request more tests
+- As a reviewer, try to make it clear to the author what conversations are
+  required changes vs. suggestions (perhaps GitHub Requested Change vs Comment?)
```
the gh api dont work here. There s no way to distinguish if you re doing both

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,34 @@
+- NQA-1 explicitly allows for a change-control board member to push commits to a
+  pull request and maintain independent reviewer status as long as the pushed
+  commits are not "significant."
+
```
```suggestion
- Reviews should be prompt. We should strive to provide a first quick review within XX business days
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,34 @@
+# Code Review
+
+## Review Guidelines
+
```
let s categorize this more
& have some more basic guidelines

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,34 @@
+  pull request and maintain independent reviewer status as long as the pushed
+  commits are not "significant."
+
+## Addressing Deadlock Between a CCB Author and CCB Reviewer
```
```suggestion
- New features should come under more scrutiny than re-factoring, which benefits from the confidence offered by the existing test suite

## Addressing Deadlock Between a CCB Author and CCB Reviewer
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,34 @@
+  commits are not "significant."
+
+## Addressing Deadlock Between a CCB Author and CCB Reviewer
+
```
maybe this should be a subsection and not its own big category. This doesnt happen often, and would happen even less with a tad more politeness

lindsayad committed:
Apply Guillaume suggestions

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

joshuahansel commented:
This file captures some of the thoughts expressed in our discussion, but I think the review page should just be a list of specific instructions, rather than giving justification for the instructions. If we wanted to still have some of the discussion captured, we could maybe have a paragraph discussing the development philosophy.

lindsayad committed:
Apply my suggestions

joshuahansel commented:
I'm not sure on the best way to iterate on this because I want to rewrite the whole page 😅 Here's my initial draft:
```
# Reviewing Pull Requests

MOOSE is a large project with large numbers of users and contributors. 
While MOOSE seeks the highest standards of software quality, it
is important to recognize that requiring perfection in proposed
code changes is not realistic, given the large amount of activity,
versus the limited availability of reviewers and time.
The guidelines outlined on this page are designed to guide
reviewers in finding the right balance between efficiency
and software quality.

## Conduct

- Be respectful.
- Try to be helpful. The reviewer is not expected to implement changes
  for the author, but the reviewer should try to ensure that
  the author has some direction, particularly if the author is a new 
  or infrequent contributor.
- Make it clear what are changes that you require, versus suggestions.
  For example, use "I suggest" for suggestions and imperative language,
  e.g., "Do this", for required changes.
- If it would be helpful, provide the author with a link to the contributing
  guidelines page.
- It is acceptable to make commits to the contributor's feature branch.
  Note that NQA-1 explicitly allows this, and an additional, independent
  review is only required if the additional commit(s) are significant.
- If the author refuses the requested changes, request an additional
  reviewer (ideally a `CODEOWNER` for the affected lines) to make a
  final decision on the changes.

## Review Strategy

- Review the high-level design of the changes before reviewing finer
  details.
- Consider the scope of the changes when determining how thorough
  the review should be: changes that potentially affect a large
  number of users should be subjected to greater scrutiny. Also
  consider the likelihood of additional capability being built
  on top of these changes.
- Carefully consider the user interface, such as input parameters,
  particularly for potentially high profile changes.
- Prioritize discovery of bugs that would manifest in very subtle,
  difficult-to-discover ways, rather than those bugs that
  would immediately become apparent, such as by an immediate error message.

## Testing

- The reviewer is not expected to test out the proposed
  changes, but this may be recommended in some cases when the 
  regression testing is deemed insufficient.
- At the very least, testing should cover most of the new code lines.
  Note that code coverage is automatically checked to meet a minimum
  value, but the reviewer may decide to override this failure if
  deemed justified. For example, a change may involve some branches
  in conditionals that should be impossible to hit.
- The reviewer may request additional testing.

## Documentation

- Inline code documentation such as doxygen is not required but may be requested
  when code is unclear.
- Request documentation for new classes and systems and also for 
  modifications to existing documentation where it is deemed important.
- Thorough documentation is not required, but it should be accurate
  and error-free, and it should clarify anything unusual or unclear.
- If the changes are significant enough to communicate to the 
  community, ensure there is a corresponding newsletter entry.

## Grammar, Spelling, and Typos

- Inline typos and grammar and spelling mistakes are recommended
  to be fixed, but are not as high of a priority as correctness in
  external documentation.
- Use Github suggestions for grammar, spelling, and typos.
```

lindsayad commented:
I would say feel free to push this document @joshuahansel. I will have some edits but I like a lot of what you have

loganharbour commented:
> I would say feel free to push this document @joshuahansel. I will have some edits but I like a lot of what you have

Just read through - I agree.

cticenhour commented:
I agree as well. Thanks @joshuahansel!

GiudGiud commented:
This is pretty much codifying a lower standard of review than I've been applying. I don't think I ll change my ways, and this should be checked for compliance with our SQA documents before we get it in. 

A few comments:
Checking out most PRs and checking that edge cases actually works is very helpful. You become the first user and find tons of bugs.
Fixing the spelling should be done as you're looking at the code, it's not that much and work and it will never get done later. 
Same for doxygen. Maybe add a note about "think about the next person trying to modify this code: is it commented enough?"
Or simply a blanket: always ask for more comments
The line coverage is easy to deceive, it s better to look at which features are being covered. 
There's also a few tentative contributors who are disrespectful / condescending, so the blanket "Be respectful" could be softened to allow some leniency. Alternatively we could have a system to close PRs for too much attitude from the tentative contributor
"- Prioritize discovery of bugs that would manifest in very subtle,
  difficult-to-discover ways, rather than those bugs that
  would immediately become apparent, such as by an immediate error message."
this is hard. What s easy is to think about to find bugs are the classic issues : higher order variables, distributed parallelism, nodal vs elemental quantities etc

Add a section about CIVET tips?
Add valgrind recipe if memory managements seems improvable
Add parallel sweeping recipes if convergence criteria are loose or the problem is inherently parallel

GiudGiud commented:
We could add something about how reviews should be prompt ?
I dont like much that some PRs get ignored for weeks because everyone is busy. Maybe we should impose it on ourselves to get to every PR for a first opinion soon?

We should also add to this the rules for assignment of reviewers and for assignees (the merge button guy)

lindsayad commented:
> This is pretty much codifying a lower standard of review than I've been applying. I don't think I ll change my ways, and this should be checked for compliance with our SQA documents before we get it in.

I'm hoping this is a document we can all eventually agree on. Let's work together to make it that way. I don't regard what @joshuahansel has written as the final version (or really anywhere close to the final version)

GiudGiud commented:
We should also add something about how the reviewer has to understand what is going on and should feel free / be encouraged to ask questions to the PR maker.

joshuahansel committed:
Updated draft of review guidelines

Refs #23580

joshuahansel commented:
Ok, done. I added this issue you can reference, by the way: https://github.com/idaholab/moose/issues/23580.

lindsayad commented:
I'm thinking I might close this

joshuahansel commented:
I agree that we need to make some kind of decision here. I think it's useful to have some kind of review guidelines page, to give suggestions if nothing else, so I think it'd be a shame to throw it all away. I have no problem removing the content telling reviewers how thorough to be, if that's the uncomfortable bit with moving forward. Thoughts @GiudGiud?

joshuahansel commented:
For example, what could definitely be kept are practical instructions about the rules and such (what situations require which reviewers, etc.). So it could give some of the information related to SQA requirements, and then additionally any practices that we'd like to impose on top of that.

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+# Reviewing Pull Requests
+
+MOOSE is a large project with large numbers of users and contributors.
+While MOOSE seeks the highest standards of software quality, it
+is important to recognize that requiring perfection in proposed
+code changes is not realistic, given the large amount of activity,
+versus the limited availability of reviewers and time.
+The guidelines outlined on this page are designed to guide
+reviewers in finding the right balance between efficiency
+and software quality.
```
```suggestion
Reviewers are bound by strict software quality standards, which they
enforce through careful reviews, often at the expense of their own research.
The guidelines outlined on this page are designed to help reviewers
provide an environment welcoming of external contributions and deal with
common situations. 
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+  for the author, but the reviewer should try to ensure that
+  the author has some direction, particularly if the author is a new
+  or infrequent contributor.
+- Make it clear what are changes that you require, versus suggestions.
+  For example, use "I suggest" for suggestions and imperative language,
+  e.g., "Do this", for required changes.
```
```suggestion
- Make it clear what are large changes that you require, versus suggestions.
  For example, use "I suggest" for suggestions and imperative language,
  e.g., "Do this", for required changes.
  Anything minor (typos, grammar, docstrings) is assumed to be required. If you
  do not wish for something to be fixed, do not suggest the change.
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+  on top of these changes.
+- Carefully consider the user interface, such as input parameters,
+  particularly for potentially high profile changes.
+- Prioritize discovery of bugs that would manifest in very subtle,
+  difficult-to-discover ways, rather than those bugs that
+  would immediately become apparent, such as by an immediate error message.
```
```suggestion
- Consider how the code will be understood outside of the context of a pull request. 
Are the comments and documentation enough to understand what is going on without
asking the initial developers.
```

lindsayad commented:
Oh yea the rule that you don't need another reviewer if you push code that "isn't significant" is definitely worth having on the books

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+
+## Conduct
+
+- Be respectful.
```
```suggestion
- Be respectful. Feel free to close the pull request if the contributor is not.
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+  e.g., "Do this", for required changes.
+- If it would be helpful, provide the author with a link to the contributing
+  guidelines page.
+- It is acceptable to make commits to the contributor's feature branch.
```
```suggestion
- Feel free to ask questions. It's normal to not understand everything in the pull request when you have
not written that code. Many times, the response to these questions should actually be added as comments
in the code.
- It is acceptable to make commits to the contributor's feature branch.
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+- The reviewer is not expected to test out the proposed
+  changes, but this may be recommended in some cases when the
+  regression testing is deemed insufficient.
+- At the very least, testing should cover most of the new code lines.
```
```suggestion
- At the very least, testing should cover all the features contributed, and most of the relevant code lines.
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+  number of users should be subjected to greater scrutiny. Also
+  consider the likelihood of additional capability being built
+  on top of these changes.
+- Carefully consider the user interface, such as input parameters,
```
```suggestion
  If the proposed changes do not allow for additional capability, this can be a reason
  to refuse them.
- Carefully consider the user interface, such as input parameters,
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+## Documentation
+
+- Inline code documentation such as doxygen is not required but may be requested
+  when code is unclear.
```
```suggestion
  when code is unclear. When unclear whether it's unclear, ask for more documentation.
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+  modifications to existing documentation where it is deemed important.
+- Thorough documentation is not required, but it should be accurate
+  and error-free, and it should clarify anything unusual or unclear.
+- If the changes are significant enough to communicate to the
```
```suggestion
- When capabilities in the code change, it's likely user documentation will need to be adapted. Always check out
  the current state of the user documentation to make sure it does not become inaccurate.
- If the changes are significant enough to communicate to the
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+  value, but the reviewer may decide to override this failure if
+  deemed justified. For example, a change may involve some branches
+  in conditionals that should be impossible to hit.
+- The reviewer may request additional testing.
```
```suggestion
- The reviewer may request additional testing.
- The reviewer may enable additional testing recipes. Anything with manual memory management
  could benefit from a valgrind recipe. Anything with loose tolerances is likely to fail parallel testing,
  feel free to add those recipes. 
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+
+## Testing
+
+- The reviewer is not expected to test out the proposed
+  changes, but this may be recommended in some cases when the
+  regression testing is deemed insufficient.
```
```suggestion
```
checking out the code locally and stress-testing it is straight up the best way to find issues. I dont think we should point the reviewer to not doing it

GiudGiud replied:
"when the
  regression testing is deemed insufficient." this is true but very hard to detect!
It s easy to get coverage while completely missing the feature even, so catching edge cases is real hard

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+  review is only required if the additional commit(s) are significant.
+- If the author refuses the requested changes, request an additional
+  reviewer (ideally a `CODEOWNER` for the affected lines) to make a
+  final decision on the changes.
```
```suggestion
  final decision on the changes.
  - Be timely. Pull requests should only linger when there's good reasons for them not to move forward.
    If you cannot review, consider delegating the review to another member.
```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,74 @@
+- If the author refuses the requested changes, request an additional
+  reviewer (ideally a `CODEOWNER` for the affected lines) to make a
+  final decision on the changes.
+
```
```suggestion
- Be a team player. If there is a section of the code that you are not familiar with,
  feel free to ask for another reviewer's opinion.

```

GiudGiud replied:
maybe this isnt the most team player advice in the end. Maybe we should find something to say we should act as a team on this though. It's important.

GiudGiud reviewed (changes requested):
If you don't want to loose your work I can understand that. Here are a few suggestions.

lindsayad commented:
> Let's debate this current set for a bit.

You mean the set of changes you just suggested?

GiudGiud commented:
yup

lindsayad committed:
Apply Guillaume's reviewing suggestions

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

joshuahansel commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,89 @@
+- Carefully consider the user interface, such as input parameters,
+  particularly for potentially high profile changes.
+- Consider how the code will be understood outside of the context of a pull request. 
+Are the comments and documentation enough to understand what is going on without
+asking the initial developers.
```
```suggestion
  The comments and documentation should be enough to understand what is going on without
  asking the initial developers.
```

joshuahansel commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,89 @@
+  in conditionals that should be impossible to hit.
+- The reviewer may request additional testing.
+- The reviewer may enable additional testing recipes. Anything with manual memory management
+  could benefit from a valgrind recipe. Anything with loose tolerances is likely to fail parallel testing,
+  feel free to add those recipes. 
```
```suggestion
  could benefit from a valgrind recipe. Anything with loose tolerances is likely to fail parallel testing.
  Feel free to add those recipes. 
```

joshuahansel commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,89 @@
+## Documentation
+
+- Inline code documentation such as doxygen is not required but may be requested
+  when code is unclear. When unclear whether it's unclear, ask for more documentation.
```
```suggestion
  when code is unclear. When in doubt, ask for more documentation.
```

joshuahansel commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,89 @@
+  modifications to existing documentation where it is deemed important.
+- Thorough documentation is not required, but it should be accurate
+  and error-free, and it should clarify anything unusual or unclear.
+- When capabilities in the code change, it's likely user documentation will need to be adapted. Always check out
+  the current state of the user documentation to make sure it does not become inaccurate.
```
```suggestion
- Ensure that existing documentation remains accurate with code changes.
```

joshuahansel reviewed (changes requested):
I'm happy with @GiudGiud's changes - I just made some minor changes.

joshuahansel committed:
Added link to review guidelines

Closes #23580

lindsayad committed:
Apply Josh reviewing guideline suggestions
+ formatting

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

Update framework/doc/content/framework/reviewing.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update framework/doc/content/framework/reviewing.md

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1458278/) on dcbe8e1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23571/site)

This comment will be updated on new commits.

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,88 @@
+The guidelines outlined on this page are designed to help reviewers
+provide an environment welcoming of external contributions and deal with
+common situations.
+
```
```suggestion

!alert note
We welcome external reviews of pull requests, even from parties with a stake in the proposed changes,
however due to the software quality requirements, pull requests will always require a review from a member
of the Change Control Board.

```

GiudGiud commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,93 @@
+provide an environment welcoming of external contributions and deal with
+common situations.
+
+!alert note
+We welcome external reviews of pull requests, even from parties with a stake in the proposed changes,
+however due to the software quality requirements, pull requests will always require a review from a member
+of the Change Control Board.
```
the "reviewer may contribute to the PR" thing is also below

also contributon to PRs are easy for maintainers, but for non-maintainers it s a process right?
either gh suggestions or make a PR to the PR on the origin fork

joshuahansel replied:
Ok, changed the suggestion. How's that?

lindsayad reviewed (commented):
I approve this message

lindsayad commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,91 @@
+- Make it clear what are large changes that you require, versus suggestions.
+  For example, use "I suggest" for suggestions and imperative language,
+  e.g., "Do this", for required changes.
+  Anything minor (typos, grammar, docstrings) is assumed to be required. If you
+  do not wish for something to be fixed, do not suggest the change.
```
```suggestion
  Anything minor (typos, grammar, docstrings) is assumed to be required. If you
  do not wish for something to be changed, do not suggest the change.
```

cticenhour commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,91 @@
+
+## Conduct
+
+- Be respectful. Feel free to close the pull request if the contributor is not.
```
```suggestion
- The MOOSE project strives to create a respectful environment for all contributors, reviewers, and users. Therefore, please be respectful during the pull request submission and approval process. Pull requests can be closed if interactions do not meet this standard. 
```

cticenhour commented on framework/doc/content/framework/reviewing.md:
```diff
@@ -0,0 +1,91 @@
+  for the author, but the reviewer should try to ensure that
+  the author has some direction, particularly if the author is a new
+  or infrequent contributor.
+- Make it clear what are large changes that you require, versus suggestions.
```
```suggestion
- Make it clear the large changes that are required by the reviewer, versus suggestions.
```

GiudGiud replied:
Clearly state required large changes ?

lindsayad committed:
Final review of the review

Co-authored-by: Casey Icenhour <cticenho@ncsu.edu>

Update framework/doc/content/framework/reviewing.md

GiudGiud commented:
only squashed some commits

lindsayad commented:
thanks team!
